### DISCUSSION TITLE
Lightweight SD card support

### DISCUSSION THREAD

**Discussion by prandeamus** [SpenceKonde/ATTinyCore#553] [General]
I'm looking at the feasibility of attaching SD card using SPI to a tiny, specifically our friend the ATTiny88. 
The standard Arduino SD package seems to be too big to fit meaningfully into 8K flash. The closest alternative I have found is elm-chan's "petit fs" at http://elm-chan.org/fsw/ff/00index_p.html which is around 2K flash and optimised to use minimal RAM buffers. The limitations of petit fs don't bother rme.

Good points
- Designed with tiny-sized flash in mind
- Configurable: so you can for example strip out the write components if you just want read-only access
- Layered design abstracts the physical layer from the FAT file system
However
- For the physical layer, elm-chan I can see only one for the ATMEGA series and one for Tiny85 which is hardwired to USI. 

It would be good to have a solution that takes advantage of the Tiny88 SPI hardware, ideally by using the TinyCore SPI abstraction. Given enough time and tantrums I could figure it out eventually, but it would be a good to stand on the shoulders of giants. Does anyone have experience of this or can point me to someone who's worked in this area?

I'm not completely wedded to the idea of SD cards, so if there are better solutions for larger storage (things like raw sound sample) I would be interested in those too.

**Comment by prandeamus** [comment]
Further to this, Elm-chan provides a generic bit-bang solution here: http://elm-chan.org/fsw/ff/pfsample.zip file in "generic" folder where there is a port bitbang driver: mmcbbp.c That will be a good place for me to start, and in theory it might even work out-of-the-box with the tiny core. That, at least, I can investigate.

**Comment by prandeamus** [comment]
More of my internal monologue, sorry,

Given that the Tiny88 does have dedicated hardware, can I assume it works the same way at register level as registers on (say) atmega328? I the comments in SPI.cpp say that Dr Azzy has basically merged the "official" Arduino SPI library and Christiansen's tinySPI. I think I'm write in saying that (as per line 23) the #ifdef SPDR and reference to hardware SPI is the place where that decision is made, I should be able to take example code for atmega328 that uses low-level access to SPI and copy/paste.

**Comment by SpenceKonde** [comment]
The ATtiny88 SPI peripheral is identical to the 328p one. The SPI peripheral is a very basic one

I thought there was some awkwardness about working with SD cards that required a 512b ram buffer though, that could be a deal breaker.

**Reply by prandeamus** [reply]
I think http://elm-chan.org/works/sd8p/report.html shows that it can be done - this is with the Tiny85 driving an SD card directly with the "petit fs" layer - but using USI to drive the SPI interface so it's very x5-specific. It uses a very small buffer, but at least for some cases it should be workable. 

Maybe there are some cards that require a 512b buffer in SPI mode - I cannot say. I think the concern is for writing: if a 512-byte write is broken into 32-byte chunks that's 16x more writes to the SD. Not only is that slower but it wears out the flash faster. 

I have the PetitFS now running on a regular Arduino Nano using the SPI library I know that can be used to basic FAT file operations on an SD card without the expense of the full SD library. If the SPI libraries in TinyCore are substantially source compatible with those in the regular Arduino I should be able to port that down to a Tiny.
